src/Evolution/Systems/NewtonianEuler/Limiters/Flattener.cpp

```diff
@@ -0,0 +1,152 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/NewtonianEuler/Limiters/Flattener.hpp"
+
+#include <array>
+#include <limits>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "NumericalAlgorithms/LinearOperators/DefiniteIntegral.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "PointwiseFunctions/Hydro/EquationsOfState/EquationOfState.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeArray.hpp"
+
+namespace NewtonianEuler::Limiters {
+
+template <size_t VolumeDim, size_t ThermodynamicDim>
+size_t flatten_solution(
+    const gsl::not_null<Scalar<DataVector>*> mass_density_cons,
+    const gsl::not_null<tnsr::I<DataVector, VolumeDim>*> momentum_density,
+    const gsl::not_null<Scalar<DataVector>*> energy_density,
+    const Mesh<VolumeDim>& mesh,
+    const Scalar<DataVector>& det_logical_to_inertial_jacobian,
+    const EquationsOfState::EquationOfState<false, ThermodynamicDim>&
+        equation_of_state) noexcept {
+  // Note on the design behind the handling of the cell-averaged fields:
+  //
+  // The cell averages are needed for
+  // - sanity-checking the inputs for validity
+  // - flattening in case of negative density
+  // - flattening in case of negative pressure
+  //
+  // In addition, we have the following optimization goals:
+  // - cell averages are only computed once
+  // - cell averages are only computed when needed
+  //
+  // To meet these goals, we create a temporary variable for each field's cell
+  // average. These temporaries live on the stack and should have minimal cost.
+  double mean_density = std::numeric_limits<double>::signaling_NaN();
+  auto mean_momentum =
+      make_array<VolumeDim>(std::numeric_limits<double>::signaling_NaN());
+  double mean_energy = std::numeric_limits<double>::signaling_NaN();
+
+  const auto compute_means = [&mean_density, &mean_momentum, &mean_energy,
+                              &mass_density_cons, &momentum_density,
+                              &energy_density, &mesh,
+                              &det_logical_to_inertial_jacobian]() noexcept {
+    // compute the means
+    const double volume_of_cell =
+        definite_integral(get(det_logical_to_inertial_jacobian), mesh);
+    const auto jacobian_weighted_mean =
+        [&mesh, &det_logical_to_inertial_jacobian,
+         &volume_of_cell](const DataVector& u) noexcept {
+          return definite_integral(get(det_logical_to_inertial_jacobian) * u,
```

Comment:
Sorry, I don't quite understand, can you clarify?

Reply:
The signature of definite integral is:
```
template <size_t Dim>
double definite_integral(const DataVector& integrand,
                         const Mesh<Dim>& mesh) noexcept;
```
so the expression `det_jac * u` does an allocation. Does that make sense?

Reply:
Understood, but I'm not sure what the alternative could be, because don't I need an allocation to store the product anyway? The only improvement I can think of would be to allocate a buffer outside the lambda and store the `det_jac * u` in there... it'd save me 3 (in 3d) allocations per call. Do you think I should do this?

Reply:
It likely won't make a difference, definitely not a big one. Maybe just add a comment to make folks aware? Expression templates make life confusing because of these subtle side effects :)